silence/silence_test.go

```diff
@@ -523,8 +520,44 @@ func TestSilenceLimits(t *testing.T) {
 	// Do not check the exact size as it can change between consecutive runs
 	// due to padding.
 	require.Contains(t, err.Error(), "silence exceeded maximum size")
-	// TODO: Once we fix https://github.com/prometheus/alertmanager/issues/3878 this should be require.Equal.
 	require.NotEqual(t, "", sil3.Id)
+
+	// Should be able to insert sil4 and then update it without modifications.
+	sil4 := &pb.Silence{
+		Matchers: []*pb.Matcher{{Name: "k", Pattern: "l"}},
+		StartsAt: time.Now(),
+		EndsAt:   time.Now().Add(5 * time.Minute),
+	}
+	require.NoError(t, s.Set(sil4))
+	require.NotEqual(t, "", sil4.Id)
+
+	// Update without modifications.
+	id4 := sil4.Id
+	require.NoError(t, s.Set(sil4))
+	require.Equal(t, id4, sil4.Id)
+
+	// Should be able to update the comment.
+	sil5 := cloneSilence(sil4)
+	sil5.Comment = "m"
+	require.NoError(t, s.Set(sil5))
+	require.Equal(t, sil4.Id, sil5.Id)
+
+	// Should not be able to update the start and end time as this requires
+	// sil5 to be expired and a new silence to be created. However, this would
+	// exceed the maximum number of silences, which counts both active and
+	// expired silences.
+	sil6 := cloneSilence(sil5)
+	sil6.StartsAt = time.Now().Add(5 * time.Minute)
+	sil6.EndsAt = time.Now().Add(10 * time.Minute)
+	err = s.Set(sil6)
+	require.Error(t, err)
+	require.EqualError(t, err, "exceeded maximum number of silences: 1 (limit: 1)")
+	require.NotEqual(t, "", sil6.Id)
+
+	// sil5 should not be expired because the update failed.
```

Comment:
This here shows that the existing silence is not expired if we are at the limit.